### PR TITLE
Add `const char *` overloads to String class

### PR DESCRIPTION
**Pull request by Rubonnek**
The added function overloads, namely:

```cpp
int _count(const char *p_string, int p_from, int p_to, bool p_case_insensitive) const;
int rfind(const char *p_str, int p_from = -1) const;
int rfindn(const char *p_str, int p_from = -1) const;
String replace_first(const char *p_key, const char *p_with) const;
String replacen(const char *p_key, const char *p_with) const;
int count(const char *p_string, int p_from = 0, int p_to = 0) const;
int countn(const char *p_string, int p_from = 0, int p_to = 0) const;
```
help avoid creating a ton of `String` objects and the underlying dynamic memory allocations they require by using the C-strings directly when possible.

I personally find that this patch makes the Editor feel snappier, but I have not run any benchmarks yet.

Edit: Added the following overloads:
```cpp
int String::findn(const char *p_str, int p_from)
bool ends_with(const char *p_string) const;
String trim_prefix(const char *p_prefix) const;
String trim_suffix(const char *p_suffix) const;
int get_slice_count(const char *p_splitter) const;
String get_slice(const char *p_splitter, int p_slice) const;
Vector<String> split(const char *p_splitter = "", bool p_allow_empty = true, int p_maxsplit = 0) const;
Vector<String> rsplit(const char *p_splitter = "", bool p_allow_empty = true, int p_maxsplit = 0) const;
```

Edit: In https://github.com/godotengine/godot/pull/84379#discussion_r1521494672 @miv391 pointed out that finding an empty string (`""`) through `String::find(const char *p_str, int p_from)` returns `0` which is inconsistent when doing the same through `GDScript`. I've updated this PR to also fix that pre existing bug since it should return `-1`.

### CODE REVIEW

File: core/string/ustring.cpp [hunk 1]
```diff
@@ -3616,6 +3790,35 @@ String String::replacen(const String &p_key, const String &p_with) const {
 	return new_string;
 }
 
+String String::replacen(const char *p_key, const char *p_with) const {
```
Comment (bruvzg) [inline core/string/ustring.cpp:3793]: Something is missing here, `p_with` is not used in the function.

**Comment by Rubonnek** [comment]
I noticed I missed another overload I wanted and took the liberty to add `int String::findn(const char *p_str, int p_from)` as well, which is the last one I'll add for this PR.

**Comment by AThousandShips** [comment]
Let's add some unit tests for these to ensure correctness, do they work with unicode? If so then ensure it works with non-utf8 input

**Comment by AThousandShips** [comment]
It currently does break current checks, and unsure what happens with various cases using non-ascii/utf8 input

File: core/string/ustring.h [hunk 2]
```diff
@@ -285,8 +286,11 @@ class String {
 	int find(const char *p_str, int p_from = 0) const; ///< return <0 if failed
 	int find_char(const char32_t &p_char, int p_from = 0) const; ///< return <0 if failed
 	int findn(const String &p_str, int p_from = 0) const; ///< return <0 if failed, case insensitive
+	int findn(const char *p_str, int p_from = -1) const; ///< return <0 if failed
```
Comment (AThousandShips) [inline core/string/ustring.h:289]:
```suggestion
	int findn(const char *p_str, int p_from = 0) const; ///< return <0 if failed
```
Incorrectly copied argument, won't work with default argument

Comment (Rubonnek) [reply core/string/ustring.h:289]: Thank you!

**Comment by Rubonnek** [comment]
> Edit: or rather, you need to rework the tests to test for the String and c-string versions, because the current tests use string literals that are now going through a different method

Indeed. I was about to ask about this -- I'm not sure if we want to duplicate the tests like this:

```cpp
	CHECK(s.findn("WHA") == 7);
	CHECK(s.findn(String("WHA")) == 7);
```

Or maybe add a macro to test both under the same line and keep it DRY.

I don't know if there's a cleaner way to do this. There doesn't seem to be.

File: tests/core/string/test_string.h [hunk 3]
```diff
@@ -348,22 +348,103 @@ TEST_CASE("[String] Substr") {
 	CHECK(s.substr(3) == "ler Baby");
 }
 
+#define MULTICHECK_STRING_EQ(obj, func, param1, eq) \
```
Comment (AThousandShips) [inline tests/core/string/test_string.h:351]:
```suggestion
#define MULTICHECK_STRING_EQ(m_obj, m_func, m_param1, m_eq) \
```
Prefix arguments in macros with `m_`

File: tests/core/string/test_string.h [hunk 4]
```diff
@@ -348,22 +348,103 @@ TEST_CASE("[String] Substr") {
...
+	do {                                            \
+		CHECK(obj.func(param1) == eq);              \
+		CHECK(obj.func(String(param1)) == eq);      \
+	} while (0)
```
Comment (AThousandShips) [inline tests/core/string/test_string.h:355]: Any reason for the `do/while`, copied from somewhere?

Comment (Rubonnek) [reply tests/core/string/test_string.h:355]: It's a common practice to help scope the macros and use them as if they were regular statements or function calls. In this particular case  I suppose I could drop them -- `CHECK` probably has them once it's expanded due to how it's used. I don't mind removing the ones I added.

Comment (AThousandShips) [reply tests/core/string/test_string.h:355]: The other checks here doesn't have them but just uses a simple empty block, I've never seen this particular practice though

Comment (Rubonnek) [reply tests/core/string/test_string.h:355]: Here's a quick example in our code base -- keep an eye on `INIT_XR_FUNC_V` and `EXT_INIT_XR_FUNC_V`:

https://github.com/godotengine/godot/blob/5ee983188de97ae027f9b9c1443438063f708a7e/modules/openxr/util.h#L36-L43


https://github.com/godotengine/godot/blob/5ee983188de97ae027f9b9c1443438063f708a7e/modules/openxr/extensions/openxr_fb_passthrough_extension_wrapper.cpp#L219-L222

Comment (Rubonnek) [reply tests/core/string/test_string.h:355]: I just tested mine without the `do/while` -- I'll drop them

**Review by AThousandShips — APPROVED** [review]
Haven't directly tested, ~and there is the possibility of code breaking with non-ascii strings passed as literals, but this applies to the already added methods like `find`, so not a new issue~, would be good to confirm that it doesn't break things like: `str.findn(U"...")`

Otherwise it looks good to me, but TIWAGOS

File: core/string/ustring.cpp [hunk 5]
```diff
@@ -3166,6 +3166,46 @@ int String::findn(const String &p_str, int p_from) const {
 	return -1;
 }
 
+int String::findn(const char *p_str, int p_from) const {
+	if (p_from < 0) {
+		return -1;
+	}
+
+	int src_len = strlen(p_str);
```
Comment (AThousandShips) [inline core/string/ustring.cpp:3174]: One concern is the universality of this, and uniformity, as `find` doesn't use it

**Comment by Rubonnek** [comment]
> would be good to confirm that it doesn't break things like: str.findn(U"...")

I should clarify I intended to keep this PR unrelated to `U""` and `L""` C-strings. Those calls won't break due to this PR because they are on an unrelated code path. The call `str.findn(U"")` yields a function signature similar to `findn(const char32_t*)`  for which there is no implementation (this PR only implements `findn(const char *)`) and as a result the compiler will look for other possible function candidates it can call given the `findn(const char32_t*)` signature. Currently, there's only one other candidate which is `str.findn(String)` which the compiler knows how to call because it can convert the `const char32_t*` parameter to a `String` through the `String` copy constructor `String::String(const char32_t*)`. The same happens with a `str.findn(L"")` function call and, again, this PR never touches either of those code paths.

Having said that, we are indeed missing some tests for `U""` and `L""` C-strings. I'll add the tests for the functions I've overloaded if they pass.

**Comment by AThousandShips** [comment]
Agreed, just wanted to make sure the automatic overload resolution works correctly as C++ can be so messy with that some times

**Comment by Rubonnek** [comment]
Tests passed for `U""` and `L""` C-strings on my end so I've added them.

File: core/string/ustring.cpp [hunk 6]
```diff
@@ -1124,6 +1124,26 @@ int String::get_slice_count(String p_splitter) const {
 	return slices;
 }
 
+int String::get_slice_count(const char *p_splitter) const {
+	if (is_empty()) {
+		return 0;
+	}
+	if (*p_splitter == '\0') {
```
Comment (AThousandShips) [inline core/string/ustring.cpp:1131]:
```suggestion
	if (p_splitter == nullptr || *p_splitter == '\0') {
```
Realized there's no check to handle null strings here

File: core/string/ustring.cpp [hunk 7]
```diff
@@ -1164,6 +1184,47 @@ String String::get_slice(String p_splitter, int p_slice) const {
 	return ""; //no find!
 }
 
+String String::get_slice(const char *p_splitter, int p_slice) const {
+	if (is_empty() || *p_splitter == '\0') {
```
Comment (AThousandShips) [inline core/string/ustring.cpp:1188]:
```suggestion
	if (is_empty() || p_splitter == nullptr || *p_splitter == '\0') {
```

File: core/string/ustring.cpp [hunk 8]
```diff
@@ -1278,6 +1339,54 @@ Vector<String> String::split(const String &p_splitter, bool p_allow_empty, int p
...
+		}
+		return ret;
+	}
+
+	int from = 0;
+	int len = length();
+
+	while (true) {
+		int end;
+		if (*p_splitter == '\0') {
```
Comment (AThousandShips) [inline core/string/ustring.cpp:1357]:
```suggestion
		if (p_splitter == nullptr || *p_splitter == '\0') {
```

**Review by AThousandShips — COMMENTED** [review]
Realized some missed details

Also needs some fixed for casting

**Comment by akien-mga** [comment]
> I personally find that this patch makes the Editor feel snappier, but I have not run any benchmarks yet.

Would be nice to do some benchmarking on an optimized build to see if there is actually a significant difference. Performance optimization can have surprising results only reasoned about but not tested.

CC @Calinou

**Comment by Calinou** [comment]
> Would be nice to do some benchmarking on an optimized build to see if there is actually a significant difference. Performance optimization can have surprising results only reasoned about but not tested.

Here benchmarks of editor startup + shutdown time (I ran these twice and results were consistent):

```
❯ hyperfine -iw1 "bin/godot.linuxbsd.editor.x86_64.master --path /tmp/4 -e --quit" "bin/godot.linuxbsd.editor.x86_64 --path /tmp/4 -e --quit"
Benchmark 1: bin/godot.linuxbsd.editor.x86_64.master --path /tmp/4 -e --quit
  Time (mean ± σ):      3.027 s ±  0.058 s    [User: 2.075 s, System: 0.361 s]
  Range (min … max):    2.911 s …  3.132 s    10 runs
 
Benchmark 2: bin/godot.linuxbsd.editor.x86_64 --path /tmp/4 -e --quit           <--- This PR
  Time (mean ± σ):      2.982 s ±  0.290 s    [User: 2.074 s, System: 0.366 s]
  Range (min … max):    2.523 s …  3.537 s    10 runs
 
Summary
  bin/godot.linuxbsd.editor.x86_64 --path /tmp/4 -e --quit ran
    1.02 ± 0.10 times faster than bin/godot.linuxbsd.editor.x86_64.master --path /tmp/4 -e --quit
```

Binary size:

### Editor

```
master: 105,335,696 bytes
    PR: 105,270,128 bytes (-65 KB)
```

### Release export template

```
master: 62,385,544 bytes
    PR: 62,344,616 bytes (-41 KB)
```

**Comment by Rubonnek** [comment]
Rebased to incorporate test changes from #89194.

File: tests/core/string/test_string.h [hunk 9]
```diff
@@ -350,18 +350,18 @@ TEST_CASE("[String] Substr") {
...
 TEST_CASE("[String] Find no case") {
 	String s = "Pretty Whale Whale";
-	CHECK(s.findn("WHA") == 7);
-	CHECK(s.findn("WHA", 9) == 13);
-	CHECK(s.findn("Revenge of the Monster SawFish") == -1);
-	CHECK(s.rfindn("WHA") == 13);
+	MULTICHECK_STRING_EQ(s, findn, "WHA", 7);
+	MULTICHECK_STRING_INT_EQ(s, findn, "WHA", 9, 13);
+	MULTICHECK_STRING_EQ(s, findn, "Revenge of the Monster SawFish", -1);
+	MULTICHECK_STRING_EQ(s, rfindn, "WHA", 13);
```
Comment (miv391) [inline tests/core/string/test_string.h:364]: How about some more tests for finds. I would also suggest separating `find` and `rfind` test to separate test cases (also `findn` and `rfindn`).

These tests are still in the old format. 

```c++
	CHECK(s.find("") == 0);
	CHECK(s.find("", 9) == 9);
	CHECK(s.find("Pretty Woman Woman") == 0);
	CHECK(s.find("WOMAN") == -1);

	CHECK(s.rfind("") == 18); // fails (returns -1)
	CHECK(s.rfind("", 15) == 15); // fails (returns -1)
	CHECK(s.rfind("foo") == -1);
	CHECK(s.rfind("Pretty Woman Woman") == 0);
	CHECK(s.rfind("man", 13) == 9);
	CHECK(s.rfind("WOMAN") == -1);

	CHECK(s.findn("") == 0); // fails (returns -1)
	CHECK(s.findn("", 3) == 3); // fails (returns -1)
	CHECK(s.findn("wha") == 7);
	CHECK(s.findn("Wha") == 7);

	CHECK(s.rfindn("") == 15); // fails (returns -1)
	CHECK(s.rfindn("", 13) == 13); // fails (returns -1)
	CHECK(s.rfindn("wha") == 13);
	CHECK(s.rfindn("Wha") == 13);
```

There are some some inconsistencies how things work in master and thus some of these tests fail. Finding with "" seems to be a somewhat undefined case. The basic `find` returns `0` when `""` is searched, which to me seems to be the correct answer. There is a `""` at every index. But all other functions (`rfind`, `findn`, `rfindn`) return `-1` when `""` is searched. In my opinion that is wrong, when `""` is searched, I would expect the first index checked to be returned.

However, in GDScript both `find` and `rfind` return `-1` when "" is searched. Although I personally would prefer not returning `-1`, consistency is more important. If it is decided that `""` is never found (-1), current C++ `find` implementation should be changed.

File: tests/core/string/test_string.h [hunk 10]
```diff
@@ -1693,9 +1676,9 @@ TEST_CASE("[String] Strip edges") {
 
 TEST_CASE("[String] Trim") {
 	String s = "aaaTestbbb";
-	CHECK(s.trim_prefix("aaa") == "Testbbb");
-	CHECK(s.trim_suffix("bbb") == "aaaTest");
-	CHECK(s.trim_suffix("Test") == s);
+	MULTICHECK_STRING_EQ(s, trim_prefix, "aaa", "Testbbb");
+	MULTICHECK_STRING_EQ(s, trim_suffix, "bbb", "aaaTest");
+	MULTICHECK_STRING_EQ(s, trim_suffix, "Test", s);
```
Comment (miv391) [inline tests/core/string/test_string.h:1681]: More tests. Again, I would suggest separate test cases for prefix and suffix.

```c++
	MULTICHECK_STRING_EQ(s, trim_prefix, "Test", s);
	MULTICHECK_STRING_EQ(s, trim_prefix, "", s);
	MULTICHECK_STRING_EQ(s, trim_prefix, "aaaTestbbb", "");
	MULTICHECK_STRING_EQ(s, trim_prefix, "bbb", s);
	MULTICHECK_STRING_EQ(s, trim_prefix, "AAA", s);
	
	MULTICHECK_STRING_EQ(s, trim_suffix, "", s);
	MULTICHECK_STRING_EQ(s, trim_suffix, "aaaTestbbb", "");
	MULTICHECK_STRING_EQ(s, trim_suffix, "aaa", s);
	MULTICHECK_STRING_EQ(s, trim_suffix, "BBB", s);
```

File: tests/core/string/test_string.h (hunk 9 above)
Comment (Rubonnek) [reply tests/core/string/test_string.h]: > The basic `find` returns `0` when `""` is searched, which to me seems to be the correct answer. There is a `""` at every index. But all other functions (`rfind`, `findn`, `rfindn`) return `-1` when `""` is searched.

That's a bug in the overload `String::find(const char *p_str, int p_from)`. It's documented it [should return `-1`](https://docs.godotengine.org/en/stable/classes/class_string.html#class-string-method-find), and the same applies to `rfind`, `findn`, and `rfindn`. The way I see it, `""` is by definition an empty string and therefore searching for it implies nothing will be found and therefore the function should return `-1`.

I've patched this bug in this PR -- `find("")` now returns `-1` to match the behavior of the rest. Thanks for pointing that out.

> separate test cases

That's a style code change better suited for a different PR. I consider that change to be out of the scope of this PR.

File: tests/core/string/test_string.h (hunk 10 above)
Comment (Rubonnek) [reply tests/core/string/test_string.h]: Tests added but kept within the same test case. There are many other test cases that could be separated such as `split`/`rsplit`, `pad`/`rpad`, and others. That codestyle change is better suited to a different PR that perhaps also adds more tests.

**Comment by akien-mga** [comment]
Seems good to me. Could you rebase to make sure it still passes CI?

**Comment by Rubonnek** [comment]
CI checks passed after rebasing.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**